standard/classes.md

```diff
@@ -5542,3 +5544,129 @@ When the body of the async function terminates, the return task is moved out of
 If the return type of the async function is `void`, evaluation differs from the above in the following way: Because no task is returned, the function instead communicates completion and exceptions to the current thread’s ***synchronization context***. The exact definition of synchronization context is implementation-dependent, but is a representation of “where” the current thread is running. The synchronization context is notified when evaluation of a `void`-returning async function commences, completes successfully, or causes an uncaught exception to be thrown.
 
 This allows the context to keep track of how many `void`-returning async functions are running under it, and to decide how to propagate exceptions coming out of them.
+
+## §indexable-sequence Indexable sequences
+
+### §indexable-sequence-general General
+
+An ***indexable sequence*** is an ordered set of zero or more elements having the same type. Any given element can be accessed via an index, and a contiguous subset of elements—referred to as a ***slice***—can be denoted via a range.
+
+An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a user-defined type is determined by the implementer of that type.
```

Comment:
Hmm. Should this be a countable type?

```suggestion
An index is represented by a read-only variable of the value type `System.Index`. A range is represented by a read-only variable of value type `System.Range`, which contains a start and end index. A slice of an array is represented by a (possibly empty) array. The representation of a slice of a countable type that is not an array type is determined by the implementer of that type.
```

Reply:
@jskeet – In C# terms a slice made from a `Range` has to be of a countable type as `Range` is built upon `Index`, and `Index` requires a length for it to be converted into an integer offset…However a non-countable type could provide the same underlying `Slice` method…

Looking at your query made me think how careful we have to be with terms here.

> An index is represented by a read-only variable of the value type `System.Index`.

No, an indexer takes an *index* as its argument, which does not need to be of type `Index`, or even `int`. So “…index **is** represented by…” is wrong, we’re using “index” to both mean a value of any type used as an index and to mean the (`int`, `bool`) pair embodied by `System.Index`…

I think we might yet need a lot of careful wordsmithing. Drat!